File: Yixuan/17_01_17/cloest_number_in_sorted_array.java
public class Solution {
    /**
     * @param A an integer array sorted in ascending order
     * @param target an integer
     * @return an integer
     * @link: http://www.lintcode.com/en/problem/closest-number-in-sorted-array/
     * @author : Egbert Li
     */
   public int closestNumber(int[] A, int target) {
      if (A.length == 0) {
         return 0;
      }
      int start = 0;
      int end = A.length - 1;
      while (start + 1 < end) {
         int mid = start + (end - start) / 2;
         if (target == A[mid]) {
            return mid;
         } else if (target < A[mid]) {
            end = mid;
         } else {
            start = mid;
         }
      }
      if (target - A[start] <= A[end] - target) {
         return start;
      }
      return end;
   }
}

File: Boxuan/16_01_17/cloest_number_in_sorted_array.py
class Solution:
    # @param {int[]} A an integer array sorted in ascending order
    # @param {int} target an integer
    # @return {int} an integer
    # @author: Egbert Li
    # @link: http://www.lintcode.com/en/problem/closest-number-in-sorted-array/
    def closestNumber(self, A, target):
        if not target:
            return -1
        left, right = 0, len(target) -1 
        while left + 1 < right:
            mid = left + (right - left) / 2
            if A[mid] == target:
                return mid
            elif A[mid] > target:
                left = mid
            else:
                right = mid
        ans = left if abs(A[left] - target) < abs(A[right] - target) else right
        return ans